### PR TITLE
[GPU] Enable multi-stream without mutex

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Enable multistream without mutex*
   - *Each network has own kernels_cache* 
   - *Separate kernels_cache::add_kernel from factory (choose_impl)*
   - *Reset kenrels_cache._kernels after kernels_cache.compile*
   - *Create cldnn unit test case to check multi-stream processing*

### Tickets:
 - *80676*

### CODE REVIEW

**Review by yeonbok — COMMENTED** [review]
Could you validate this this on the entire benchmark?

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 1]
```diff
@@ -213,6 +214,13 @@ struct network {
     /// Returns memory state @p variable_id of stateful network
     VariableState& get_variable_memory(const std::string &variable_id);
 
+    /// Add APIs for handling kernels_cache
+    void compile();
+    kernel_id add_kernel(const std::shared_ptr<kernel_string>& kernel_sring);
+    kernel::ptr get_kernel(kernel_id id);
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp:220]: I think we can mark this method as `const`

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 2]
```diff
@@ -289,21 +289,17 @@ void primitive_inst::update_impl() {
...
             _impl = _node.type()->choose_impl(_node, *_impl_params);
-            bool lru_popped = cache.add(layout_key, _impl->clone());
-            if (lru_popped) {
-                for (auto& id : lru->get_kernel_ids())
-                    _network.get_program()->remove_kernel(id);
-            }
-            _network.get_program()->compile();
+            _impl->add_kernel_codes(get_network().get_kernels_cache());
+            get_network().compile();
+            _impl->init_kernels(get_network().get_kernels_cache());
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:300]: I'd change the interfaces to something like that:
```cpp
auto& kernels_cache = get_network().get_kernels_cache();
kernels_cache.add(_impl->get_kernels_data());
kernels_cache.compile();
_impl->init_kernels(kernels_cache);
kernels_cache.reset();
```
or even that:
```cpp
auto& kernels_cache = get_network().get_kernels_cache();
auto kernels = kernels_cache.compile_kernels(_impl->get_kernels_data());
_impl->set_kernels(kernels);
```

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:300]: @vladimir-paramuzov I think the second is better. When do you merge this code change to master?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:300]: I rephrased the comment: I **propose** to change this part in a way above. So please implement it within this PR.

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp:223]: I will change it. thanks.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:300]: Okay, I see. I will update the code . thanks.

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Done.

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp]: Remove these APIs because they are not necessary any more after refactoring codes.

**Comment by ahnyoung-paul** [comment]
@yeonbok , @vladimir-paramuzov , @sshlyapn Could you please review again?

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp [hunk 3]
```diff
@@ -249,5 +259,14 @@ struct network {
     void check_names();
     void add_default_output_chains();
     output_chains_map::iterator add_output_chain(std::shared_ptr<primitive_inst>& p_inst);
+
+    std::unique_ptr<kernels_cache> _kernels_cache;
+    // TODO: Move from cldnn::program to cldnn::network for multi-thread issue.
+    // Need to consider the location of these caches for more optimizations.
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp:265]: I think this comment is redundant at this point.

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 4]
```diff
@@ -325,6 +326,11 @@ network::network(program::ptr program, stream::ptr stream, bool is_internal, boo
         wait_for_the_turn();
     }
 
+    _kernels_cache = std::unique_ptr<kernels_cache>(new kernels_cache(program->get_engine(), program->get_id(),
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/network.cpp:329]: I think we only need to have this when the network is dynamic?

File: src/plugins/intel_gpu/src/runtime/kernels_cache.cpp [hunk 5]
```diff
@@ -384,4 +384,42 @@ void kernels_cache::reset() {
     _pending_compilation = false;
 }
 
+std::vector<kernel_id> kernels_cache::add(std::vector<std::shared_ptr<kernel_string>> kernel_sources, bool dump_custom_program) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/runtime/kernels_cache.cpp:387]: can we rename the function more illustratively, add_kernels_source?

File: src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp (hunk 3 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/graph/network.hpp]: Done

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 4 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/network.cpp]: Done

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: @vladimir-paramuzov Could you please review it again?

**Review by e-ddykim — APPROVED** [review]
Looks good to me.

nit. I think it could improve readability of code if KernelsCache or kernels_cache is renamed properly.

**Comment by ahnyoung-paul** [comment]
> LGTM for the implementation itself. However , since it is modifying the common codes, could you validate this this on the entire benchmark?

I run dlbenchmark for all omz models, not found regression.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**